zulip/zulip/__init__.py

```diff
@@ -93,9 +95,11 @@ def fail(self):
         super(RandomExponentialBackoff, self).fail()
         # Exponential growth with ratio sqrt(2); compute random delay
         # between x and 2x where x is growing exponentially
-        delay_scale = int(2 ** (self.number_of_retries / 2.0 - 1)) + 1
-        delay = delay_scale + random.randint(1, delay_scale)
-        message = "Sleeping for %ss [max %s] before retrying." % (delay, delay_scale * 2)
+        delay_base = 0.5
+        delay = random.random() * self.delay_cap
+        if math.log2(self.delay_cap) > self.number_of_retries:
```

Comment:
Ugh, I think my prior suggestion was a bit misleading.  I didn't want us to actually calculate the logarithm at run time; I was just explaining mathematically that after log2N delays, we're always gonna get `delay_cap` here instead of `2 ** retries`, but I didn't mean to actually code it this way.

It's my fault here for probably prematurely optimizing.  (Well, I wasn't prematurely optimizing, I was more mistakenly worried about overflow bugs.)  I think JS can handle large exponents of 2.

Reply:
@showell changed the code accordingly. The optimization could be added later if needed.